File: CodeForces/2001-/2132 - Codeforces Round 1043 (Div. 3)/C2. The Cunning Seller (hard version).cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

int main() {
    int t;
    cin >> t;
    
    const long long INF = 1e18;
    vector<long long> cost = {3, 10, 33, 108, 351, 1134, 3645, 11664, 37179, 118098, 373977, 1180980, 3720087, 11691702, 36669429, 114791256, 358722675, 1119214746, 3486784401, 10847773692, 33705582543};
    vector<long long> number = {1, 3, 9, 27, 81, 243, 729, 2187, 6561, 19683, 59049, 177147, 531441, 1594323, 4782969, 14348907, 43046721, 129140163, 387420489, 1162261467, 3486784401};
    
    while (t--) {
        int n, k;
        cin >> n >> k;
        
        vector<vector<long long>> dp(k + 1, vector<long long>(n + 1, INF));
        dp[0][0] = 0;
        
        for (int i = 1; i <= k; i++) {
            for (int j = 0; j <= n; j++) {
                dp[i][j] = dp[i - 1][j];
                for (int x = 0; x < cost.size(); x++) {
                    if (number[x] <= j) {
                        dp[i][j] = min(dp[i][j], dp[i - 1][j - number[x]] + cost[x]);
                    }
                }
            }
        }
        
        if (dp[k][n] != INF) {
            cout << dp[k][n] << endl;
        } else {
            cout << -1 << endl;
        }
    }
    
    return 0;
}

File: CodeForces/2001-/2132 - Codeforces Round 1043 (Div. 3)/C2. The Cunning Seller (hard version).py
t = int(input())
INF = 10**18
cost = [3, 10, 33, 108, 351, 1134, 3645, 11664, 37179, 118098, 373977, 1180980, 3720087, 11691702, 36669429, 114791256, 358722675, 1119214746, 3486784401, 10847773692, 33705582543]
number = [1, 3, 9, 27, 81, 243, 729, 2187, 6561, 19683, 59049, 177147, 531441, 1594323, 4782969, 14348907, 43046721, 129140163, 387420489, 1162261467, 3486784401]


for _ in range(t):
    n, k = map(int, input().split())

    if k >= 30:
        temp_n = n
        deals = 0
        for i in range(len(number) - 1, -1, -1):
            x = temp_n // number[i]
            temp_n %= number[i]
            deals += x

        if deals <= k:
            temp_n = n
            coin = 0
            for i in range(len(number) - 1, -1, -1):
                x = temp_n // number[i]
                temp_n %= number[i]
                coin += x * cost[i]
            print(coin)
            continue

    max_reachable = min(n, k * max(number[:10]))

    INF = float('inf')
    dp = {}
    dp[(0, 0)] = 0

    for i in range(1, k + 1):
        new_dp = {}
        for (prev_deals, watermelons), prev_cost in dp.items():
            if prev_deals == i - 1:
                key = (i, watermelons)
                if key not in new_dp:
                    new_dp[key] = INF
                new_dp[key] = min(new_dp[key], prev_cost)

                for x in range(min(15, len(cost))):
                    new_watermelons = watermelons + number[x]
                    if new_watermelons <= n:
                        key = (i, new_watermelons)
                        if key not in new_dp:
                            new_dp[key] = INF
                        new_dp[key] = min(new_dp[key], prev_cost + cost[x])

        dp.update(new_dp)

    result = INF
    for deals in range(k + 1):
        key = (deals, n)
        if key in dp:
            result = min(result, dp[key])

    if result != INF:
        print(result)
    else:
        print(-1)